themes/default/content/blog/iac-recommended-practices-using-stack-references/index.md

```diff
@@ -0,0 +1,132 @@
+
+1. When it comes to stack outputs, **export what is needed.** If it needs to be accessed from outside the stack, export it; otherwise, don't. Why? Stack outputs form the application programming interface (API) by which other Pulumi programs interact with this Pulumi program. Like in any API implementation, expose what the user cares about or is likely to need, and leave the rest as an implementation detail. You can add more stack outputs after the fact with only minimal code changes, and as has been mentioned already this is done with a quick `pulumi up` that has no affect on existing infrastructure (you're modifying the stack object itself).
+2. **Be judicious in your use of stack references.** This is sort of the "flip side" of the previous bullet. If you need to reference an output from another stack, then do so; otherwise, don't. If you find yourself needing lots of stack references, then the way in which you've structured your projects may need to be re-examined so as to minimize the coupling between projects. Additionally, there is a potential "speed of light" concern here; if where you are executing your Pulumi code is across a high-latency link from where you store your Pulumi state, stack references will be affected by that latency.
+3. **If you need to expose lots of outputs on a stack, consider exporting a structured data object.** It's possible to construct a JSON object (or dict or struct, depending on your language) to hold all the stack outputs, and then export that object. Then a stack reference can read that object, resulting in a single call to the backend. Be aware, though, that you'll need to write the necessary code to understand/import/unmarshall that JSON object in the referring stack, so there is a small amount of extra work required in this situation.
+4. In line with parameterizing as much of your code as possible, **also be sure to parameterize your stack references.** A stack reference is built using an organization name, a project name, and a stack name. Don't hardcode these values; instead, use configuration values to allow the users to specify from which source stack(s) the values will be referenced. You'll see an example of that in the next section.
+5. **Plan for handling cascading stack updates.** If a stack has outputs being consumed via stack references, what happens when those stack outputs change? The Pulumi Service [supports webhooks](/docs/intro/pulumi-service/webhooks/) that can help, and Pulumi is currently exploring options for having Pulumi Deployments help with this (see [this GitHub issue](https://github.com/pulumi/service-requests/issues/181)). In practice this should be uncommon, and frequent occurrences may be a good indicator that your resource partitioning (how resources are split among multiple projects) needs to be adjusted.
+
```

Comment:
I would add a 6th bullet point talking about security. With separate stacks also comes that not everyone should be allowed to access the stack outputs of the production stack. I would highlight [Stack Permissions](https://www.pulumi.com/docs/intro/pulumi-service/projects-and-stacks/#stack-permissions) and [Teams and RBAC](https://www.pulumi.com/docs/intro/pulumi-service/teams/) to control such access.

Reply:
This is an excellent suggestion as a revision to the previous post, which discusses _why_ you would use multiple projects. Unless I'm misreading your suggestion, it sounds like it would fall squarely in that post as opposed to this one (which is focused on the use of stack references).

Reply:
I've created https://github.com/pulumi/devrel-team/issues/530 to track this work.

Reply:
I would integrate it here. After splitting the teams (article 3), you are now introducing the stack references. If you layer projects (infra, k8s, app), you can also talk about stack permissions explaining that the access to the infra and/or k8s is read-only to the app team, while read-write for the infra team.

If you export secrets in a stack, you definitely want to limit access to these, hence also my suggestion (in this article) to narrow the permissions overall to the production stacks of each Pulumi program.

Reply:
Keep in mind that the next post in the series (after we take a quick break to look at local testing) is security and RBAC, in which we will most definitely look at this in more detail. I'll see if I can work in a quick mention here with a statement that we'll be exploring the topic in more detail in an upcoming post.

Reply:
Fixed in b34f9cd9